### PR TITLE
Dev eeprom userspace

### LINKED ISSUE #248: Mem_address nomenclature
mem_address is used for every function, this may be misleading. Use proper naming for the variable

### PR DESCRIPTION
**Pull request by sjpark608**
Closes #248, #244, #151, #249

line 184 to 308 in edgepi_eeprom.py are added for user space memory access.

### CODE REVIEW

File: src/edgepi/calibration/edgepi_eeprom.py [hunk 1]
```diff
@@ -171,3 +181,128 @@ def page_write_register(self, mem_addr: int = None, data: list = None):
...
+        Read user space memory starting from 0 to 16383
+        Args:
+            mem_addr (int): starting memory address to read from
+            length (int): length of data to read
+        Return:
+            data (list): list of data read from the specified memory and length
+        """
+        self.__parameter_sanity_check(start_addrx, length, True)
+        dummy_data = self.__generate_list_of_pages(start_addrx, [0]*length)
+        start_addrx = start_addrx + EdgePiMemoryInfo.USER_SPACE_START_BYTE.value
```
Comment (naomiven) [inline src/edgepi/calibration/edgepi_eeprom.py:240]: why is there an x at the end of `start_addrx`

File: src/edgepi/calibration/edgepi_eeprom.py [hunk 2]
```diff
@@ -171,3 +181,128 @@ def page_write_register(self, mem_addr: int = None, data: list = None):
...
+        address_offset = 0
+        data_read = []
+        for data in dummy_data:
```
Comment (naomiven) [inline src/edgepi/calibration/edgepi_eeprom.py:243]: why is this called `dummy_data` ? is it actual dummy data?

File: src/edgepi/calibration/edgepi_eeprom.py [hunk 3]
```diff
@@ -171,3 +181,128 @@ def page_write_register(self, mem_addr: int = None, data: list = None):
...
+        else:
+            is_full=False
+            is_empty=False
+            mem_start = EdgePiMemoryInfo.USER_SPACE_START_BYTE.value + \
+                        EdgePiMemoryInfo.BUFF_START.value
+            # read memory content should be in json encoded bytes converted into list
+            mem_content = bytes(self.read_memory(mem_start, mem_size))
+            self.data_list = json.loads(mem_content)
+            self.used_size = mem_size
+            self.log.info(f'{mem_size}bytes of data is read from the user space')
```
Comment (naomiven) [inline src/edgepi/calibration/edgepi_eeprom.py:306]: why is `is_empty` True if `mem_size` is 0xFFFF? is `mem_size` how much memory is taken? it looks like a big number to me so i thought it should be NOT empty

File: src/test_edgepi/unit_tests/test_calibration/test_access_eeprom.py [hunk 4]
```diff
@@ -21,6 +24,13 @@ def fixture_test_dac(mocker):
     mocker.patch("edgepi.peripherals.i2c.I2C")
     yield EdgePiEEPROM()
 
+# Max data size of osensa space
+OSENSA_DATA_SIZE = EdgePiMemoryInfo.USER_SPACE_START_BYTE.value-3
```
Comment (naomiven) [inline src/test_edgepi/unit_tests/test_calibration/test_access_eeprom.py:28]: what is 3 :)

**Review by naomiven — COMMENTED** [review]
looks good.. a bit confused about some parts

File: src/edgepi/calibration/edgepi_eeprom.py (hunk 1 above)
Comment (sjpark608) [reply src/edgepi/calibration/edgepi_eeprom.py:240]: didn't want to spell out the address but wanted it to sound similar 😅

File: src/edgepi/calibration/edgepi_eeprom.py (hunk 2 above)
Comment (sjpark608) [reply src/edgepi/calibration/edgepi_eeprom.py:243]: yes, the I2C reading operation requires a dummy list of the same length as the data you want to read from other device

File: src/edgepi/calibration/edgepi_eeprom.py (hunk 3 above)
Comment (sjpark608) [reply src/edgepi/calibration/edgepi_eeprom.py:306]: I'm using the first two bytes of user space memory to represents the size of memory in use. When the memory is reset, all values in the memory are set to 0xFF which results in the mem_size being 0xFFFF

File: src/test_edgepi/unit_tests/test_calibration/test_access_eeprom.py (hunk 4 above)
Comment (sjpark608) [reply src/test_edgepi/unit_tests/test_calibration/test_access_eeprom.py:28]: Osensa reserved space also uses the first two bytes to represents the used data. `OSENSA_DATA_SIZE` is a dummy data size to test memory bound edge case. 
size of `EdgePiMemoryInfo.USER_SPACE_START_BYTE.value-3` should fit in Osensa reserved space with a room for 1 more byte. Adding 1 more byte makes the memory space full. Adding 2 more bytes raises the MemoryOutOfBound Error

**Review by farzadpanahi — APPROVED** [review]
